src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/io_context.h"
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+/// C++11 helper for generalized lambda capture.
+template <typename T, typename F>
+class capture_impl {
+  T x;
+  F f;
+
+ public:
+  capture_impl(T &&x, F &&f) : x{std::forward<T>(x)}, f{std::forward<F>(f)} {}
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+};
+
+template <typename T, typename F>
+capture_impl<T, F> capture(T &&x, F &&f) {
+  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+}
+
+/// Set stats collection flag based on corresponding Ray config flag.
+const bool io_context_proxy::stats_collection_enabled_ =
+    RayConfig::instance().asio_event_loop_stats_collection_enabled();
+
+void io_context_proxy::post(std::function<void()> handler, const std::string &name) {
+  if (!stats_collection_enabled_) {
+    return boost::asio::io_context::post(std::move(handler));
+  }
+  {
+    absl::WriterMutexLock lock(&(global_stats_.mutex));
+    global_stats_.stats.cum_count++;
+    global_stats_.stats.curr_count++;
+  }
+  // Get this handler's stats.
+  mutex_.ReaderLock();
+  auto it = post_handler_stats_.find(name);
+  mutex_.ReaderUnlock();
+  if (it == post_handler_stats_.end()) {
+    // Lock the table until we have added the entry. We use try_emplace and handle a
+    // failed insertion in case the item was added before we acquire the writers lock;
+    // this allows the common path, in which the handler already exists in the hash table,
+    // to only require the readers lock.
+    absl::WriterMutexLock lock(&mutex_);
+    const auto pair =
+        post_handler_stats_.try_emplace(name, std::make_shared<GuardedHandlerStats>());
+    if (pair.second) {
+      it = pair.first;
+    } else {
+      it = post_handler_stats_.find(name);
+      // If try_emplace failed to insert the item, the item is guaranteed to exist in
+      // the table.
+      RAY_CHECK(it != post_handler_stats_.end());
+    }
+  }
+  {
+    absl::WriterMutexLock lock(&(it->second->mutex));
+    it->second->stats.cum_count++;
+    it->second->stats.curr_count++;
+  }
+  const auto start = std::chrono::steady_clock::now();
```

Comment:
Can you use absl::GetCurrentTimeNanos() instead of this?

Reply:
Oh neat, I didn't know that we had standardized around that.